crates/sui-source-validation-service/src/lib.rs

```diff
@@ -197,18 +221,59 @@ pub async fn verify_packages(
         }
     }
 
+    let mut lookup = BTreeMap::new();
     for t in tasks {
-        t.await.unwrap()?;
+        let new_lookup = t.await.unwrap()?;
+        lookup.extend(new_lookup);
     }
-    Ok(())
+    Ok(lookup)
+}
+
+pub struct AppState {
+    pub sources: SourceLookup,
 }
 
-pub fn serve() -> anyhow::Result<Server<AddrIncoming, IntoMakeService<Router>>> {
-    let app = Router::new().route("/api", get(api_route));
+pub fn serve(app_state: AppState) -> anyhow::Result<Server<AddrIncoming, IntoMakeService<Router>>> {
+    let app = Router::new().route("/api", get(api_route).with_state(Arc::new(app_state)));
     let listener = TcpListener::bind("0.0.0.0:8000")?;
     Ok(Server::from_tcp(listener)?.serve(app.into_make_service()))
 }
 
-async fn api_route() -> &'static str {
-    "{\"source\": \"code\"}"
+#[derive(Deserialize)]
+pub struct Request {
+    address: String,
+    module: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SourceResponse {
+    pub source: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ErrorResponse {
+    pub error: String,
+}
+
+async fn api_route(
+    State(app_state): State<Arc<AppState>>,
+    Query(Request { address, module }): Query<Request>,
+) -> impl IntoResponse {
+    let symbol = Symbol::from(module);
+    let Ok(address) = AccountAddress::from_hex_literal(&address) else {
+	let error = format!("Invalid hex address {address}");
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
+    };
+    let Some(path) = app_state.sources.get(&(address, symbol)) else {
+	let error = format!("No source found for {symbol} at address {address}" );
+	return (StatusCode::BAD_REQUEST, Json(ErrorResponse { error }).into_response())
+    };
+    let Ok(source) = fs::read_to_string(path) else {
```

Comment:
We may want to cache these in memory rather than read them from disk for every request (and in future employ a more elaborate caching mechanism) -- even though we don't have many packages, I expect people might look at the system packages fairly often.

Reply:
`+1` going to iterate on this